### PR TITLE
sensu: nixos module - WIP

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

This PR introduces full support for the sensu monitoring platform.

This is still very much WIP but we are currently using this in production on 17.03.

I will finalize this for 17.09.

###### Things done

Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers.

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/monitoring/sensu/default.nix [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+      then (map (s: cleanAttrSet s) set_or_list)
+      else
+        lib.mapAttrsRecursive (path: v:
+          (if builtins.isString v then (lib.replaceStrings ["\n"] [""] v) else v))
+          (lib.filterAttrsRecursive (k: v:
+            !(k == "_module" || v == "" || v == [] || v == {})) set_or_list);
+
+  checkCfg = pkgs.writeText "checks.json" (builtins.toJSON (cleanAttrSet {
+    inherit (cfg) checks;
+  }));
```
Comment (Mic92) [inline nixos/modules/services/monitoring/sensu/default.nix]: There are quite a few options, you have added here. Do you think you can maintain it in future versions that or would it be better to relax the checks and just convert attributes to JSON.

**Comment by peterhoeg** [comment]
A quick comment to @cstrahan and @Mic92 (+ whoever else may have taken an interest in this) - please do not waste your time going through this in detail as it is still being changed.

If you find any glaring bits, feel free to shout of course.

**Comment by mmahut** [comment]
Any update on this pull request?

**Comment by peterhoeg** [comment]
We are using sensu *extensively* based on this PR (with a lot of custom bits that are not part of this), but as sensu legacy is being phased out in favour of the go version, this will require extensive changes. 

Those changes we will have to do sometime before sensu legacy is retired (end this year).

**Pull request closed without merging**